### ISSUE TITLE
Design Questions 

### ISSUE THREAD

**Issue by bluesky314** [NathanUA/U-2-Net#27]
Thanks for you amazing work. I learnt alot from your 2 papers. I had a few questions:

1) Why have you only used Cross entropy loss and not also ssim and iou as in basnet? How are advantages of those losses, which were outlined with detailed analysis in basnet, made up for with only CE here?

2) Why did you set all the deep-supervision weights to 1? It is normal to set them to values between 0.2-0.8 so the model focuses most on the final output.

3) There does not seem to be a lr scheduler which is a default today. May i know why you did not use one?

4) How much did taking all the predications and concatenating them to predict the final map make a difference vs just taking the top most prediction? As a 1X1 conv is used, we are just taking a linear combination of the previous prediction and it is apparent that the later most predictions will be the most accurate. Did you look at the weights learnt by the 1X1 to see what was going on?

**Comment by NathanUA** [comment]
1. Why have you only used Cross entropy loss and not also ssim and iou as in basnet? How are advantages of those losses, which were outlined with detailed analysis in basnet, made up for with only CE here?
RES: This is a good question. Because the design of U2Net enables it to achieve different level features from the RSU regardless of the feature maps' resolution and the multi-level hybrid losses doesn't improve that much in this case. The BSI loss in basnet actually costs more time than CE. So we don't use the bsi loss in U2Net. 

Why did you set all the deep-supervision weights to 1? It is normal to set them to values between 0.2-0.8 so the model focuses most on the final output. 
RES: Almost all of the settings are based on experiences. There is no clear evidence which setting is the best. But we acknowledge that our setting is not the optimal. If you want , you can try grid search to achieve better settings. 

There does not seem to be a lr scheduler which is a default today. May i know why you did not use one?
RES: We think adam is enough for us. Another reason is we want the training process to be plain so that the results can be reproduced by others easily. 

How much did taking all the predications and concatenating them to predict the final map make a difference vs just taking the top most prediction? As a 1X1 conv is used, we are just taking a linear combination of the previous prediction and it is apparent that the later most predictions will be the most accurate. Did you look at the weights learnt by the 1X1 to see what was going on?
RES: The idea actually comes from HED. The difference sometimes is not that much but sometimes make differences. Applying 1x1 convolution contains the possibility of weighting the side outputs with 0 and the last output as 1, if the last output is the most accurate. The accuracy varies from different tasks. We want the model to be general in different tasks including region segmentation and edge detection, etc. So we keep that fusion layer just like HED.

**Comment by vincentzhang** [comment]
To add to Nathan's answer, we believe that less is more. That has been the guiding design principle. We try to keep only the core contributions in the paper. The user can always optimize the algorithm further.
